Cheat Engine

RipleyTom opened the pull request:
First draft of a cheat manager.

Note that the cheats you make probably won't survive another launch of the game if the offset is on the heap. I may add a more complex parameter to make some level of redirection possible in the future.

edit:
I added a redirection system, completely untested, syntax is:
$ is offset
[] for dereferencing
+/- for adding offset
ex: [[$]+8]+16
It will get the u32 offset is pointing at, add 8 to that, get the u32 that is pointing to and add 16 to that which will be the final offset.

If the Script cell is left blank it will act as before.

Megamouse commented on rpcs3/rpcs3qt/main_window.cpp:
```diff
@@ -1331,6 +1332,11 @@ void main_window::CreateConnects()
 		trop_manager->show();
 	});
 
+	connect(ui->actionManage_Cheats, &QAction::triggered, [=] {
```
newline. (I'm starting to like it like this though... work, you know XD)

Megamouse commented on rpcs3/rpcs3qt/cheat_manager.cpp:
```diff
@@ -0,0 +1,660 @@
+
+	// Edit/Manage UI
+	connect(tbl_cheats, &QTableWidget::itemClicked, [=](QTableWidgetItem* item) {
+		int row           = tbl_cheats->currentRow();
```
row can be -1 if you click on empty space

RipleyTom replied:
I put a check just in case.

Megamouse commented on rpcs3/rpcs3qt/cheat_manager.cpp:
```diff
@@ -0,0 +1,660 @@
+	});
+
+	connect(tbl_cheats, &QTableWidget::cellChanged, [=](int row, int column) {
+		if (column != 1)
```
also check for row maybe

RipleyTom replied:
This is only called when a cell has been edited(after the edit) so I think it's ok like that?

Megamouse commented on rpcs3/rpcs3qt/cheat_manager.cpp:
```diff
@@ -0,0 +1,660 @@
+	}
+	}
+
+	if (!res_conv)
```
```suggestion
	if (res_conv)
	{
		offsets_found = g_cheat.search(to_search, offsets_found);
	}
	else
```

RipleyTom replied:
I can't do that because search is a template and I need to specify or let the compiler deduce the template type.

Megamouse commented on rpcs3/rpcs3qt/cheat_manager.cpp:
```diff
@@ -0,0 +1,660 @@
+
+	connect(lst_search, &QListWidget::customContextMenuRequested, [=](const QPoint& loc) {
+		QPoint globalPos      = lst_search->mapToGlobal(loc);
+		QMenu* menu           = new QMenu();
```
put menu below return

Megamouse commented on rpcs3/rpcs3qt/cheat_manager.cpp:
```diff
@@ -0,0 +1,660 @@
+					res_apply = g_cheat.set_value(cheat->offset, to_set);
+				break;
+			}
+			}
```
default missing

RipleyTom commented:
Fixed a bug where saves wouldn't be loaded properly if script was empty.
Cleaned up the code(no more duplicate code, const functions are now const and static functions are now static).
I tried to implement a function to try to find a pointer to address back in "safe" memory(data segment) but from testing it doesn't seem to work very well so I disabled it but left the code in, in case anyone comes up with an improvement.

elad335 commented on rpcs3/rpcs3qt/cheat_manager.cpp:
```diff
@@ -0,0 +1,859 @@
+	if (Emu.IsRunning())
+	{
+		did_we_pause = true;
+		Emu.Pause();
```
This is unsafe, use `cpu_thread::suspend_all cpu_lock(nullptr);` to suspend execution of all threads. (scope based)

elad335 replied:
It works regardless if the emu is paused or not.

elad335 commented on rpcs3/rpcs3qt/cheat_manager.cpp:
```diff
@@ -0,0 +1,840 @@
+		// Looks through mapped memory
+		for (u32 page_ind = (0x10000 / 4096); page_ind < (0xF0000000 / 4096); page_ind++)
+		{
+			if (vm::check_addr(page_ind * 4096, 1, vm::page_info_t::page_allocated | vm::page_info_t::page_readable))
```
I know you are aware of this, leaving it here as a reminder.
```suggestion
			if (vm::check_addr(page_ind * 4096))
```

elad335 commented on rpcs3/rpcs3qt/cheat_manager.cpp:
```diff
@@ -0,0 +1,840 @@
+	{
+		for (const auto& off : to_filter)
+		{
+			if (vm::check_addr(off, 1, vm::page_info_t::page_allocated | vm::page_info_t::page_readable))
```
Here as well.

elad335 commented on rpcs3/rpcs3qt/cheat_manager.cpp:
```diff
@@ -0,0 +1,840 @@
+template <typename T>
+bool cheat_engine::set_value(const u32 offset, const T value)
+{
+	if (!vm::check_addr(offset, vm::page_info_t::page_allocated | vm::page_info_t::page_writable))
```
Missing a cpu_lock here?

RipleyTom replied:
Good catch!

Nekotekina commented on rpcs3/rpcs3qt/cheat_manager.cpp:
```diff
@@ -0,0 +1,860 @@
+		result = str.toUShort(&success);
+
+	if constexpr (std::is_same<T, u32>::value)
+		result = str.toULong(&success);
```
toUInt

Nekotekina commented on rpcs3/rpcs3qt/cheat_manager.cpp:
```diff
@@ -0,0 +1,860 @@
+	if constexpr (std::is_same<T, s16>::value)
+		result = str.toShort(&success);
+
+	if constexpr (std::is_same<T, s32>::value)
```
toInt
